[Unity] Add memory scope and nd allocation support in allocators

quic-sanirudh opened the pull request:
This patch adds a new Allocator type called NDAllocator as discussed in [this discussion](https://discuss.tvm.apache.org/t/memory-scope-for-vm-alloc-storage-builtins/15172).

The new allocator allows allocating to other memory scopes and allows nd-allocation.

quic-sanirudh commented:
cc @tqchen

quic-sanirudh commented:
With this PR, the `alloc_storage`'s `size` parameter seems to have 2 different meanings based on which allocator we use. In the existing allocators, size is always expected to be 1d and it means the actual bytes of storage. With nd-allocator, it represents the physical shape of the buffer, and size is calculated from that.

This mostly came up because of the AllocDataSpace API doing a similar thing, but I think we might need to make it both represent shape, or both represent the size of allocation. In case of 2d, it'll be size of allocation for each physical dimension.

tqchen commented on include/tvm/runtime/relax_vm/memory_manager.h:
```diff
@@ -71,6 +72,14 @@ class Allocator {
+   * \param mem_scope The memory scope of allocated tensor.
+   *  \return A sized allocation in the form of a buffer.
+   */
+  virtual Buffer Alloc(int ndim, const int64_t* shape, DLDataType type_hint, String mem_scope);
```
Let us change int ndim, const int64_t* shape to ShapeTuple

quic-sanirudh replied:
Thanks, I've updated the API with `ShapeTuple`.

quic-sanirudh replied:
Actually I just realized that we could probably unify the 2 Alloc overloads into one and call the appropriate AllocDataSpace based on the size of the ShapeTuple. Was that intended change mentioned here?

tqchen replied:
let us keep things in two for now as there might be some future thoughts on byte reuse, as the original alloca is dtype agnostic

tqchen commented on include/tvm/runtime/relax_vm/memory_manager.h:
```diff
@@ -71,6 +72,13 @@ class Allocator {
+   * \param mem_scope The memory scope of allocated tensor.
+   *  \return A sized allocation in the form of a buffer.
+   */
+  virtual Buffer Alloc(ShapeTuple shape, DLDataType type_hint, String mem_scope);
```
AllocDataSpaceWithScope to be consistent with the original one

tqchen replied:
In this case type_hint should just be dtype

quic-sanirudh replied:
right makes sense, I'll fix that, thanks.

quic-sanirudh replied:
Done thanks

tqchen commented on src/runtime/relax_vm/nd_allocator.h:
```diff
@@ -0,0 +1,86 @@
+
+  Buffer Alloc(size_t nbytes, size_t alignment, DLDataType type_hint) override {
+    Buffer buf;
+    buf.device = device_;
```
seems we don't need ND Allocator, just can let Naive allocator redirects to device API. We might need some better strategy later for things like hexagon allocator incase caching or pooling logic is nedded

tqchen replied:
Sorry, i was meant to comment on the nd version's `Alloc`. Seems that we can have both 1D and ND directly implemented in the NaiveAllocator atm.

quic-sanirudh replied:
ah okay, yeah that makes sense. I'll get rid of the NDAllocator and implement ND version with NaiveAllocator.

quic-sanirudh replied:
One problem I see though is that the difference in how `alloc_storage` with scope (which takes physical shape) differs from existing `alloc_storage` (that takes raw bytes) could cause confusion. Especially when we get an alloc_storage call with 1d shape, and global memory scope, do we consider the 1d shape as raw bytes, or just physical shape (number of elements)?

I think unifying this discrepancy one way or another is needed if we have to get rid of NDAllocator.

quic-sanirudh replied:
I've removed `NDAllocator`. Thanks for the feedback.

tqchen commented:
Thanks @quic-sanirudh , left some extra comments. Agree about your point about allocator convention. One idea is always use the AllocWithScope API and in the case of raw bytes, pass in dtype as the element.

tqchen commented:
I actually mean that we can update `allocate_storage` to always take in shape instead of nbytes. We can always `allocate_storage` with dtype='uint8' that matches the older behavior. Then at least from VM API level the confusion is addressed, and we can continue to propagate that below

quic-sanirudh commented:
> I actually mean that we can update `allocate_storage` to always take in shape instead of nbytes. We can always `allocate_storage` with dtype='uint8' that matches the older behavior. Then at least from VM API level the confusion is addressed, and we can continue to propagate that below

Thanks for the clarification. That makes sense. I'll update the PR that way.

quic-sanirudh commented:
I've updated the PR to change alloc_storage to take shape. Right now, the `VMBuiltinLower` is updated so that `vm.alloc_storage` always allocates with size as shape and dtype as `uint8`. One thing to note is that AllocDataSpace takes a type_hint argument, but as far as I can tell, only opengl runtime needed that and so for now, passing in a different dtype to AllocDataSpace isn't going to hurt any backends, but we may want to fix it.

I'm planning to do that in a follow-up PR where I change the `relax.memory.alloc_storage` builtin generated by StaticPlanBlockMemory pass is also updated to take in shape and generates it with proper dtype. This would also allow me to later extend it to perhaps do the memory planning on a different storage_scope and probably NDAllocation.

tqchen commented:
Thanks @quic-sanirudh , this is now merged

tqchen commented:
also cc @MasterJH5574 @srkreddy1238

quic-sanirudh commented:
> Thanks @quic-sanirudh , this is now merged

Thanks a lot for all the help.
